src/main/java/de/epiceric/shopchest/utils/ShopUtils.java

```diff
@@ -278,120 +288,128 @@ public void updateShops(Player player, boolean force) {
+                    } else {
+                        shop.getItem().hidePlayer(player);
+                    }
+                }
 
-        Shop shop = getShop(loc);
-        if (shop != null) {
-            shops.add(shop);
-        } else if ((shop = getShop(below)) != null) {
-            shops.add(shop);
-        }
+                // Display hologram based on sight
+                if (shop.hasHologram()) {
+                    if (distanceSquared < hologramDistSquared) {
 
-        return shops;
-    }
+                        Location targetLocation = shop.getHologram().getLocation();
+                        targetLocation.setY(shop.getLocation().getY() + 1);
 
-    /**
-     * Update hologram and item of the shop for a player based on their distance to each other
-     * @param shop Shop to update
-     * @param player Player to show the update
-     */
-    public void updateShop(Shop shop, Player player) {
-        double holoDistSqr = Math.pow(plugin.getShopChestConfig().maximal_distance, 2);
-        double itemDistSqr = Math.pow(plugin.getShopChestConfig().maximal_item_distance, 2);
+                        double angle = targetLocation.toVector().subtract(pVec).angle(pDir);
+                        double distance = Math.sqrt(distanceSquared);
+
+                        // Check if is targeted
+                        if (angle * distance < TARGET_THRESHOLD) {
+                            // Display even if not the nearest
+                            if (!firstShopInSight) {
+                                shop.getHologram().showPlayer(player);
+                                continue;
+                            }
+
+                            if (nearestShop == null) {
+                                // nearestShop is null
+                                // => we guess this one will be the nearest
+                                nearestShop = shop;
+                                nearestDistance = distance;
+                                continue;
+                            } else if (distance < nearestDistance) {
+                                // nearestShop is NOT null && this shop is nearest
+                                // => we'll hide nearestShop, and guess this one will be the nearest
+                                otherShopsInSight.add(nearestShop);
+                                nearestShop = shop;
+                                nearestDistance = distance;
+                                continue;
+                            }
+                            // else: hologram is farther than nearest, so we hide it
+                        }
+                    }
+
+                    // If not in sight
+                    shop.getHologram().hidePlayer(player);
+                }
+            }
+        }
+
+        for (Shop shop : otherShopsInSight) {
+            // we already checked hasHologram() before adding it
+            shop.getHologram().hidePlayer(player);
```

Comment:
The other shops in sight should only be hidden if ``only-show-first-shop-in-sight`` is enabled.